binary-data-formats.qmd

```diff
@@ -0,0 +1,236 @@
+
+```{r glimpse-data}
+#| cache: true
+glimpse(seattle_lib)
+```
+
+The first line in the output tell you that `seattle_lib` is stored locally on-disk as a single csv file.
+The second line indicates that the data are stored as a tabular data set with 41 million rows and 12 columns.
+
+### Partitioning datasets
+
+When data sets become large it is often useful to split them across many files rather than storing everything in a single large file.
+When this structuring is done intelligently, the multi-file strategy can lead to significant improvements in performance because some analyses will only require a subset of the files to be read.
+
+When deciding to pursue a multi-file strategy, the first task is to decide how to partition your data set.
+If you break the data set into too few files you won't gain the performance improvements you're hoping for, but if you break it into too many you won't gain much either.
+If you split the data in a meaningful way, you'll get some performance advantages, but if you split the data randomly you probably won't see a lot of gains.
+There are no hard and fast rules about how to partition a data set: the results will depend on your data, access patterns, and the systems that read the data.
+As a rough guide, the arrow package documentation suggests the following heuristics:
+
+-   Avoid files smaller than 20MB and larger than 2GB.
+-   Avoid partitioning layouts that produce more than 10,000 distinct files.
+
+As a general rule, a useful strategy for partitioning a large dataset is to create meaningful subsets based on one or more variables that you're likely to use to filter the data.
+
+### Writing datasets
+
+The guidelines in the last section suggest that the Seattle library data is a little too large to store as a single file.
+How should we split it up?
+It's reasonable to guess that our analyses might filter the data by `CheckoutYear`, so we'll use that as our partitioning variable.
+
+Partitioning a data set can be done using the `write_dataset()` function from the arrow package, which respects the grouping created by the dplyr `group_by()` function.
+To create a multi-file data set that contains a separate file for each year of library checkout data, we can do this:
+
+```{r write-dataset}
+#| eval: !expr "!file.exists('data/seattle-library-checkouts/CheckoutYear=2021/part-0.parquet')"
+seattle_lib |>
+  group_by(CheckoutYear) |>
+  write_dataset(
+    path = "data/seattle-library-checkouts",
+    format = "parquet"
+  )
+```
+
+The `path` argument is used to specify the folder into which all the data files will be written, and the `format` argument is used to specify the file format for the data files.
+We'll talk more about the parquet format in @sec-parquet.
+For the moment, let's take a look at the files:
+
+```{r show-parquet-files}
+seattle_lib_files <- list.files(
+  path = "data/seattle-library-checkouts",
+  recursive = TRUE
+) 
+seattle_lib_files
```

Comment:
I wonder if it's worth turning this into a tibble? Then you could have a column for file size (in MB?) and it would automatically truncate to just the first six files.

Reply:
good idea!